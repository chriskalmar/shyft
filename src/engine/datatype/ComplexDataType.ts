export class ComplexDataType {}

export const isComplexDataType = (obj: unknown): obj is ComplexDataType => {
  return obj instanceof ComplexDataType;
};
