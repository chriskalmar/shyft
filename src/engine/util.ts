import * as casual from 'casual';
import * as _ from 'lodash';
import { viewAttributePropertiesWhitelist } from './constants';

type StringFunction = () => string;

export const deleteUndefinedProps = (obj) =>
  Object.keys(obj).forEach((key) => obj[key] === undefined && delete obj[key]);

export const passOrThrow = (
  condition: any,
  messageFn: string | StringFunction,
): void => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function');
  }
  if (!condition) {
    throw new Error(messageFn());
  }
};

export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : { ...functionOrMap };
};

export const isMap = (map: object, nonEmpty?: boolean): boolean => {
  return (
    map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0))
  );
};

export const isFunction = (fn: any): boolean => {
  return typeof fn === 'function';
};

export const isArray = (set: any[], nonEmpty?: boolean): boolean => {
  return (
    set !== null &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0))
  );
};

export const isString = (str: any): boolean => typeof str === 'string';

export const mergeMaps = (first, second) => {
  if (!isMap(first) || !isMap(second)) {
    throw new Error('mergeMaps() expects 2 maps for a merge to work');
  }

  return { ...first, ...second };
};

export const mapOverProperties = (
  object: object,
  iteratee: (val: any, key: any) => any,
) => {
  passOrThrow(isMap(object), () => 'Provided object is not a map');

  passOrThrow(
    isFunction(iteratee),
    () => 'Provided iteratee is not a function',
  );

  const keys = Object.keys(object);

  keys.forEach((key) => {
    iteratee(object[key], key);
  });
};

export const sortDataByKeys = (
  keys: string[],
  data: object[],
  keyProperty = 'id',
) => {
  const map = {};
  const result = [];
  const order = {};

  if (!keys || (isArray(keys) && keys.length === 0)) {
    return [];
  }

  if (!data || (isArray(data) && data.length === 0)) {
    return keys.map(() => null);
  }

  for (let d = 0; d < data.length; d++) {
    const row = data[d];
    const id = row[keyProperty];
    const found = map[id];

    if (!found) {
      map[id] = row;
    } else if (found instanceof Array) {
      found.push(row);
    } else {
      map[id] = [found, row];
    }
  }

  for (let k = 0; k < keys.length; k++) {
    const key = keys[k];
    const found = map[key];

    if (typeof found === 'undefined') {
      result.push(null);
    } else if (found instanceof Array) {
      let idx = order[key];

      if (typeof idx === 'undefined') {
        idx = 0;
      } else {
        idx++;
      }

      order[key] = idx;

      result.push(found[idx]);
    } else {
      result.push(found);
    }
  }

  return result;
};

export const reverseString = (str) => {
  return str.split('').reverse().join('');
};

export const randomJson = () => {
  const ret = {};
  const count = casual.integer(1, 5);

  _.times(count, () => {
    const key = _.camelCase(casual.words(2));
    const value = Math.random() > 0.5 ? casual.title : casual.integer();

    ret[key] = value;
  });

  return ret;
};

export const asyncForEach = async (
  array: any[],
  callback: (val, idx, arr) => any,
) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array);
  }
};

export const isDefined = (val: any): boolean => typeof val !== 'undefined';

export const convertEntityToViewAttribute = (attribute) => {
  if (!_.isObject(attribute)) {
    throw new Error(
      'convertEntityToViewAttribute() expects an attribute as input',
    );
  }

  const newAttribute = {};

  viewAttributePropertiesWhitelist.map((prop) => {
    newAttribute[prop] = attribute[prop];
  });

  return newAttribute;
};

export const convertEntityToViewAttributesMap = (attributesMap) => {
  const newAttributesMap = {};

  if (!_.isObject(attributesMap)) {
    throw new Error(
      'convertEntityToViewAttributesMap() expects an attributes map as input',
    );
  }

  for (const [name, attribute] of Object.entries(attributesMap)) {
    newAttributesMap[name] = convertEntityToViewAttribute(attribute);
  }

  return newAttributesMap;
};
