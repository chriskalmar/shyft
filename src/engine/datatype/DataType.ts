import { Entity, ShadowEntity, ViewEntity } from '../..';
import { passOrThrow, isFunction } from '../util';
import { ComplexDataType } from './ComplexDataType';

export type DataTypeSetup = {
  name?: string;
  description?: string;
  mock?: () => any;
  validate?: () => any;
  enforceRequired?: boolean;
  defaultValue?: () => any;
  enforceIndex?: boolean;
};

export type DataTypeFunction = (params: {
  setup: DataTypeSetup;
  entity?: Entity | ShadowEntity | ViewEntity;
}) => DataType | ComplexDataType;

export class DataType {
  name: string;
  description: string;
  mock?: () => any;
  validator?: (value: any, context: any) => any;
  enforceRequired?: boolean;
  defaultValue?: () => any;
  enforceIndex?: boolean;

  constructor(setup: DataTypeSetup = {} as DataTypeSetup) {
    const {
      name,
      description,
      mock,
      validate,
      enforceRequired,
      defaultValue,
      enforceIndex,
    } = setup;

    passOrThrow(name, () => 'Missing data type name');
    passOrThrow(
      description,
      () => `Missing description for data type '${name}'`,
    );

    passOrThrow(
      isFunction(mock),
      () => `'Missing mock function for data type '${name}'`,
    );

    if (validate) {
      passOrThrow(
        isFunction(validate),
        () => `'Invalid validate function for data type '${name}'`,
      );

      this.validator = validate;
    }

    if (defaultValue) {
      passOrThrow(
        isFunction(defaultValue),
        () => `'Invalid defaultValue function for data type '${name}'`,
      );

      this.defaultValue = defaultValue;
    }

    this.name = name;
    this.description = description;
    this.mock = mock;

    if (enforceRequired) {
      this.enforceRequired = enforceRequired;
    }

    if (enforceIndex) {
      this.enforceIndex = enforceIndex;
    }
  }

  validate = async (value: any, context: any): Promise<void> => {
    if (value && this.validator) {
      await this.validator(value, context);
    }
  };

  toString(): string {
    return this.name;
  }
}

export const isDataType = (obj: unknown): obj is DataType => {
  return obj instanceof DataType;
};
