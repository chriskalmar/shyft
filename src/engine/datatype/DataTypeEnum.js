import { passOrThrow, isArray } from '../util';

import { enumValueRegex, ENUM_VALUE_PATTERN } from '../constants';

import DataType from './DataType';

export class DataTypeEnum extends DataType {
  constructor(setup = {}) {
    const { name, description, values } = setup;

    passOrThrow(
      isArray(values, true),
      () => `'Missing enum values for data type '${name}'`,
    );

    values.map(value => {
      passOrThrow(
        enumValueRegex.test(value),
        () =>
          `Invalid enum value '${value}' for data type '${name}' (Regex: /${ENUM_VALUE_PATTERN}/)`,
      );
    });

    super({
      ...setup,
      description: description || `Enumeration set: ${values.join(', ')}`,
      mock() {
        const randomPos = Math.floor(Math.random() * values.length);
        return values[randomPos];
      },
    });

    this.values = values;
  }

  toString() {
    return this.name;
  }
}

export const isDataTypeEnum = obj => {
  return obj instanceof DataTypeEnum;
};
