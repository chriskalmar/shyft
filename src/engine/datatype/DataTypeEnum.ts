import { uniq } from 'lodash';
import { passOrThrow, isMap } from '../util';
import { enumValueRegex, ENUM_VALUE_PATTERN } from '../constants';
import { DataType } from './DataType';

export type ValueType = {
  [key: string]: number;
};

export type DataTypeEnumSetupType = {
  name: string;
  description?: string;
  values: ValueType;
};

export class DataTypeEnum extends DataType {
  values: ValueType;

  constructor(setup: DataTypeEnumSetupType = {} as DataTypeEnumSetupType) {
    const { name, description, values } = setup;

    passOrThrow(
      isMap(values, true),
      () => `'Missing enum values for data type '${name}'`,
    );

    const valueNames = Object.keys(values);
    const uniqueIds = [];

    valueNames.map((valueName) => {
      const valueId = values[valueName];
      uniqueIds.push(valueId);

      passOrThrow(
        enumValueRegex.test(valueName),
        () =>
          `Invalid value name '${valueName}' for data type '${name}' (Regex: /${ENUM_VALUE_PATTERN}/)`,
      );

      passOrThrow(
        valueId === Number(valueId) && valueId > 0,
        () =>
          `Value '${valueName}' for data type '${name}' has an invalid unique ID (needs to be a positive integer)`,
      );
    });

    passOrThrow(
      uniqueIds.length === uniq(uniqueIds).length,
      () =>
        `Each value defined for data type '${name}' needs to have a unique ID`,
    );

    super({
      name,
      description: description || `Enumeration set: ${valueNames.join(', ')}`,
      enforceIndex: true,
      mock() {
        const randomPos = Math.floor(Math.random() * uniqueIds.length);
        return uniqueIds[randomPos];
      },
    });

    this.values = values;
  }

  toString(): string {
    return this.name;
  }
}

export const isDataTypeEnum = (obj: any): boolean => {
  return obj instanceof DataTypeEnum;
};
