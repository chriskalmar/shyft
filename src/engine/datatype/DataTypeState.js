import { passOrThrow, isMap } from '../util';

import { stateNameRegex, STATE_NAME_PATTERN } from '../constants';

import * as _ from 'lodash';

import DataType from './DataType';

export class DataTypeState extends DataType {
  constructor(setup = {}) {
    const { name, description, states } = setup;

    passOrThrow(
      isMap(states, true),
      () => `'Missing states for data type '${name}'`,
    );

    const stateNames = Object.keys(states);
    const uniqueIds = [];

    stateNames.map(stateName => {
      const stateId = states[stateName];
      uniqueIds.push(stateId);

      passOrThrow(
        stateNameRegex.test(stateName),
        () =>
          `Invalid state name '${stateName}' for data type '${name}' (Regex: /${STATE_NAME_PATTERN}/)`,
      );

      passOrThrow(
        stateId === parseInt(stateId, 10) && stateId > 0,
        () =>
          `State '${stateName}' for data type '${name}' has an invalid unique ID (needs to be a positive integer)`,
      );
    });

    passOrThrow(
      uniqueIds.length === _.uniq(uniqueIds).length,
      () =>
        `Each state defined for data type '${name}' needs to have a unique ID`,
    );

    super({
      ...setup,
      description: description || `States: ${stateNames.join(', ')}`,
      enforceIndex: true,
      mock() {
        const randomPos = Math.floor(Math.random() * uniqueIds.length);
        return uniqueIds[randomPos];
      },
    });

    this.states = states;
  }

  toString() {
    return this.name;
  }
}

export const isDataTypeState = obj => {
  return obj instanceof DataTypeState;
};
