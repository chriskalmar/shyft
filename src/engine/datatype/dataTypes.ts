import { DataType } from './DataType';
import { DataTypeUser } from './DataTypeUser';
import { randomJson } from '../util';
import * as casual from 'casual';

export const DataTypeUserID = new DataTypeUser({
  name: 'DataTypeUserID',
  description: 'Data type representing a reference to a user',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeID = new DataType({
  name: 'DataTypeID',
  description: 'Data type representing unique IDs',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeInteger = new DataType({
  name: 'DataTypeInteger',
  description: 'Data type representing integer values',
  /* istanbul ignore next */
  mock: () => casual.integer(-2 ^ 10, 2 ^ 10),
});

export const DataTypeBigInt = new DataType({
  name: 'DataTypeBigInt',
  description: 'Data type representing big integer values',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeFloat = new DataType({
  name: 'DataTypeFloat',
  description: 'Data type representing float values',
  /* istanbul ignore next */
  mock: () => casual.double(-2 ^ 10, 2 ^ 10),
});

export const DataTypeDouble = new DataType({
  name: 'DataTypeDouble',
  description: 'Data type representing double precision values',
  /* istanbul ignore next */
  mock: () => casual.double(-2 ^ 10, 2 ^ 10),
});

export const DataTypeBoolean = new DataType({
  name: 'DataTypeBoolean',
  description: 'Data type representing boolean values',
  /* istanbul ignore next */
  mock: () => casual.boolean,
  enforceRequired: true,
  defaultValue() {
    return false;
  },
});

export const DataTypeString = new DataType({
  name: 'DataTypeString',
  description: 'Data type representing text values',
  /* istanbul ignore next */
  mock: () => casual.title,
});

export const DataTypeJson = new DataType({
  name: 'DataTypeJson',
  description: 'Data type representing json objects',
  /* istanbul ignore next */
  mock: randomJson,
});

export const DataTypeTimestamp = new DataType({
  name: 'DataTypeTimestamp',
  description: 'Data type representing a timestamp',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTimestampTz = new DataType({
  name: 'DataTypeTimestampTz',
  description: 'Data type representing a timestamp with time zone information',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeDate = new DataType({
  name: 'DataTypeDate',
  description: 'Data type representing a date',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTime = new DataType({
  name: 'DataTypeTime',
  description: 'Data type representing time',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTimeTz = new DataType({
  name: 'DataTypeTimeTz',
  description: 'Data type representing time with time zone information',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeUUID = new DataType({
  name: 'DataTypeUUID',
  description: 'Data type representing a UUID',
  /* istanbul ignore next */
  mock: () => casual.uuid,
});

export const DataTypeI18n = new DataType({
  name: 'DataTypeI18n',
  description: 'Data type representing translation objects',
  /* istanbul ignore next */
  mock: randomJson,
});

export const DataTypeUpload = new DataType({
  name: 'DataTypeUpload',
  description: 'Data type representing a file upload',
  mock: () => null,
});
