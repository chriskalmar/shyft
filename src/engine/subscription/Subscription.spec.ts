/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/explicit-function-return-type */

import {
  Subscription,
  isSubscription,
  SUBSCRIPTION_TYPE_CREATE,
  SUBSCRIPTION_TYPE_UPDATE,
  SUBSCRIPTION_TYPE_DELETE,
  processEntitySubscriptions,
  pubsub,
} from './Subscription';
import { Entity } from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import { passOrThrow } from '../util';
import { generateTestSchema } from '../../graphqlProtocol/test-helper';
import { generateGraphQLSchema } from '../../graphqlProtocol/generator';
import { subscribe, parse } from 'graphql';

describe('Subscription', () => {
  const entity = new Entity({
    name: 'SomeEntityName',
    description: 'Just some description',
    attributes: {
      someAttribute: {
        type: DataTypeString,
        description: 'Just some description',
      },
      anotherAttribute: {
        type: DataTypeString,
        description: 'Just some description',
      },
    },
  });

  it('should have a name', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Subscription();
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have a type', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Subscription({
        name: 'example',
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have a valid type', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Subscription({
        name: 'example',
        type: 12346,
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have a description', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Subscription({
        name: 'example',
        type: SUBSCRIPTION_TYPE_CREATE,
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have a list of default attributes', () => {
    const subscription = new Subscription({
      name: 'example',
      type: SUBSCRIPTION_TYPE_CREATE,
      description: 'subscribe the world',
    });

    processEntitySubscriptions(entity, [subscription]);
    const defaultAttributes = subscription.attributes;

    const expectedAttributes = ['someAttribute', 'anotherAttribute'];

    expect(defaultAttributes).toEqual(expectedAttributes);
  });

  it('should have a list of valid attribute names', () => {
    const subscription = new Subscription({
      name: 'example',
      type: SUBSCRIPTION_TYPE_CREATE,
      description: 'subscribe the world',
      attributes: ['anything', { foo: 'bar' }],
    });

    function fn() {
      processEntitySubscriptions(entity, [subscription]);
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should allow an empty attributes list for UPDATE type subscriptions', () => {
    const subscription = new Subscription({
      name: 'example',
      type: SUBSCRIPTION_TYPE_UPDATE,
      description: 'subscribe the world',
      attributes: [],
    });

    processEntitySubscriptions(entity, [subscription]);
    expect(subscription.attributes).toEqual([]);
  });

  it('should allow an empty attributes list for DELETE type subscriptions', () => {
    const subscription = new Subscription({
      name: 'example',
      type: SUBSCRIPTION_TYPE_DELETE,
      description: 'subscribe the world',
      attributes: [],
    });

    processEntitySubscriptions(entity, [subscription]);
    expect(subscription.attributes).not.toBeDefined();
  });

  it('should have a list of unique attribute names', () => {
    const subscription = new Subscription({
      name: 'example',
      type: SUBSCRIPTION_TYPE_CREATE,
      description: 'subscribe the world',
      attributes: ['anything', 'anything'],
    });

    function fn() {
      processEntitySubscriptions(entity, [subscription]);
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it("should return it's name", () => {
    const subscription = new Subscription({
      name: 'example',
      type: SUBSCRIPTION_TYPE_UPDATE,
      description: 'mutate the world',
      attributes: ['anything'],
    });

    expect(subscription.name).toBe('example');
    expect(String(subscription)).toBe('example');
  });

  it('should have a valid preProcessor function', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Subscription({
        name: 'example',
        type: SUBSCRIPTION_TYPE_CREATE,
        description: 'mutate the world',
        attributes: ['anything'],
        preProcessor: 'not-a-function',
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have a valid postProcessor function', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Subscription({
        name: 'example',
        type: SUBSCRIPTION_TYPE_CREATE,
        description: 'mutate the world',
        attributes: ['anything'],
        postProcessor: 'not-a-function',
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  describe('isSubscription', () => {
    const subscription = new Subscription({
      name: 'example',
      type: SUBSCRIPTION_TYPE_UPDATE,
      description: 'mutate the world',
      attributes: ['anything'],
      preProcessor() {
        return null;
      },
      postProcessor() {
        return {};
      },
    });

    it('should recognize objects of type Subscription', () => {
      function fn() {
        passOrThrow(
          isSubscription(subscription),
          () => 'This error will never happen',
        );
      }

      expect(fn).not.toThrow();
    });

    it('should recognize non-Subscription objects', () => {
      function fn() {
        passOrThrow(
          isSubscription({}) ||
            isSubscription(function test() {}) ||
            isSubscription(Error),
          () => 'Not a Subscription object',
        );
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('processEntitySubscriptions', () => {
    it('should throw if provided with an invalid list of subscriptions', () => {
      const subscriptions = {
        foo: [{}],
      };

      function fn() {
        processEntitySubscriptions(entity, subscriptions);
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should throw if provided with an invalid subscription', () => {
      const subscriptions = [{ foo: 'bar' }];

      function fn() {
        processEntitySubscriptions(entity, subscriptions);
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    // this check seems useless for subscription
    it.skip('should throw if required attribute is missing in CREATE type subscriptions', () => {
      function fn() {
        const otherEntity = new Entity({
          name: 'SomeEntityName',
          description: 'Just some description',
          attributes: {
            someAttribute: {
              type: DataTypeString,
              description: 'Just some description',
            },
            neededAttribute: {
              type: DataTypeString,
              description: 'This is important',
              required: true,
            },
          },
          subscriptions: [
            new Subscription({
              type: SUBSCRIPTION_TYPE_CREATE,
              name: 'onBuild',
              description: 'build item',
              attributes: ['someAttribute'],
            }),
          ],
        });

        otherEntity.getSubscriptionByName('onBuild');
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should throw on duplicate subscription names', () => {
      const subscriptions = [
        new Subscription({
          type: SUBSCRIPTION_TYPE_CREATE,
          name: 'onBuild',
          description: 'build item',
          attributes: ['someAttribute'],
        }),
        new Subscription({
          type: SUBSCRIPTION_TYPE_DELETE,
          name: 'onBuild',
          description: 'build item',
          attributes: ['someAttribute'],
        }),
      ];

      function fn() {
        processEntitySubscriptions(entity, subscriptions);
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should throw if unknown attributes are used', () => {
      const subscriptions = [
        new Subscription({
          type: SUBSCRIPTION_TYPE_CREATE,
          name: 'onBuild',
          description: 'build item',
          attributes: ['doesNotExist'],
        }),
      ];

      function fn() {
        processEntitySubscriptions(entity, subscriptions);
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should allow for empty attribute lists on DELETE type subscriptions', () => {
      const subscriptions = [
        new Subscription({
          type: SUBSCRIPTION_TYPE_DELETE,
          name: 'onDrop',
          description: 'drop item',
          attributes: [],
        }),
      ];

      processEntitySubscriptions(entity, subscriptions);
    });
  });

  describe('preProcessor', () => {
    const testEntity = new Entity({
      name: 'SomeTestsEntityName',
      description: 'Just some description',
      attributes: {
        someAttribute: {
          type: DataTypeString,
          description: 'Just some description',
          required: false,
        },
        anotherAttribute: {
          type: DataTypeString,
          description: 'Just some description',
        },
      },
      subscriptions: [
        new Subscription({
          name: 'SomeSubWithPreProcessor',
          type: SUBSCRIPTION_TYPE_CREATE,
          description: 'build item',
          attributes: ['someAttribute'],
          delimiter: '/',
          // wildCard: '',
          // pattern: '',
          preProcessor: (
            _entity,
            _id,
            _source,
            input,
            typeName,
            entitySubscription,
          ) => {
            if (entitySubscription.attributes && Object.keys(input).length) {
              const delimiter = entitySubscription.delimiter;
              const filled = entitySubscription.attributes
                .map(attribute => input[attribute])
                .reduce((acc, curr) => `${acc + delimiter + curr}`, '');

              const topic = `${entitySubscription.name}${_entity.name}${filled}`;
              return topic;
            }

            return null;
          },
        }),
      ],
    });

    const someAttribute = 'test';

    it('should pass through preProcessor if it is declared', async () => {
      const subscriptionByName = testEntity.getSubscriptionByName(
        'SomeSubWithPreProcessor',
      );
      expect(subscriptionByName).toMatchSnapshot();

      const setup = await generateTestSchema({ entities: [testEntity] });
      const graphqlSchema = generateGraphQLSchema(setup.configuration);
      const subscriptionDoc = parse(`
        subscription someSubWithPreProcessorSomeTestsEntityName($input: SomeSubWithPreProcessorSomeTestsEntityNameInput!) {
          someSubWithPreProcessorSomeTestsEntityName(input: $input) {
            someTestsEntityName {
              # someAttribute
              anotherAttribute
            }
          }
        }`);

      // subscription topic will be automatically generated by shyft
      let subscription = (await subscribe({
        schema: graphqlSchema,
        document: subscriptionDoc,
        variableValues: {
          input: {
            someTestsEntityName: {},
          },
        },
        contextValue: { pubsub },
      })) as AsyncIterableIterator<any>;

      let pending = subscription.next();

      await pubsub.publish('SomeSubWithPreProcessorSomeTestsEntityName', {
        // someTestsEntityName: {
        someAttribute,
        anotherAttribute: 'world',
        // },
      });

      let result = await pending;
      expect(result).toMatchSnapshot('withoutPreProcessorResult');
      expect(await subscription.return()).toMatchSnapshot(
        'withoutPreProcessorEnd',
      );

      // subscription topic will be generated in preProcessor
      subscription = (await subscribe({
        schema: graphqlSchema,
        document: subscriptionDoc,
        variableValues: {
          input: {
            someTestsEntityName: {
              someAttribute,
            },
          },
        },
        contextValue: { pubsub },
      })) as AsyncIterableIterator<any>;

      pending = subscription.next();

      await pubsub.publish(
        `SomeSubWithPreProcessorSomeTestsEntityName/${someAttribute}`,
        {
          someAttribute,
          anotherAttribute: 'world',
        },
      );

      result = await pending;
      expect(result).toMatchSnapshot('withPreProcessorResult');
      expect(await subscription.return()).toMatchSnapshot(
        'withPreProcessorEnd',
      );
    });
  });

  describe('postProcessor', () => {
    const testEntity = new Entity({
      name: 'SomeTestsEntityName',
      description: 'Just some description',
      attributes: {
        someAttribute: {
          type: DataTypeString,
          description: 'Just some description',
          required: false,
        },
        anotherAttribute: {
          type: DataTypeString,
          description: 'Just some description',
        },
      },
      subscriptions: [
        new Subscription({
          name: 'SomeSubWithPostProcessor',
          type: SUBSCRIPTION_TYPE_CREATE,
          description: 'build item',
          attributes: ['someAttribute'],
          delimiter: '/',
          postProcessor: (
            _entity,
            // _id,
            _source,
            input,
            typeName,
            entitySubscription,
            context,
          ) => {
            if (context && context.changePayload) {
              return { anotherAttribute: 'earth' };
            }
            return null;
          },
        }),
      ],
    });

    const someAttribute = 'test';

    it('should pass through postProcessor if it is declared', async () => {
      const subscriptionByName = testEntity.getSubscriptionByName(
        'SomeSubWithPostProcessor',
      );
      expect(subscriptionByName).toMatchSnapshot();

      const setup = await generateTestSchema({ entities: [testEntity] });
      const graphqlSchema = generateGraphQLSchema(setup.configuration);
      const subscriptionDoc = parse(`
        subscription someSubWithPostProcessorSomeTestsEntityName($input: SomeSubWithPostProcessorSomeTestsEntityNameInput!) {
          someSubWithPostProcessorSomeTestsEntityName(input: $input) {
            someTestsEntityName {
              # someAttribute
              anotherAttribute
            }
          }
        }`);

      let subscription = (await subscribe({
        schema: graphqlSchema,
        document: subscriptionDoc,
        variableValues: {
          input: {
            someTestsEntityName: {},
          },
        },
        contextValue: { pubsub },
      })) as AsyncIterableIterator<any>;

      let pending = subscription.next();

      await pubsub.publish('SomeSubWithPostProcessorSomeTestsEntityName', {
        someAttribute,
        anotherAttribute: 'world',
      });

      let result = await pending;
      expect(result).toMatchSnapshot('withoutPostProcessorResult');
      expect(await subscription.return()).toMatchSnapshot(
        'withoutPostProcessorEnd',
      );

      subscription = (await subscribe({
        schema: graphqlSchema,
        document: subscriptionDoc,
        variableValues: {
          input: {
            someTestsEntityName: {},
          },
        },
        contextValue: { pubsub, changePayload: true },
      })) as AsyncIterableIterator<any>;

      pending = subscription.next();

      await pubsub.publish('SomeSubWithPostProcessorSomeTestsEntityName', {
        someAttribute,
        anotherAttribute: 'world',
      });

      result = await pending;
      expect(result).toMatchSnapshot('withPostProcessorResult');
      expect(await subscription.return()).toMatchSnapshot(
        'withPostProcessorEnd',
      );
    });
  });
});
