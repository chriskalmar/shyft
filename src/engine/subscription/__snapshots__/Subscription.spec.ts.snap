// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscription isSubscription should recognize non-Subscription objects 1`] = `"Not a Subscription object"`;

exports[`Subscription postProcessor should pass through postProcessor if it is declared 1`] = `
Subscription {
  "attributes": Array [
    "someAttribute",
  ],
  "delimiter": "/",
  "description": "build item",
  "ignoreRequired": true,
  "isTypeCreate": true,
  "name": "SomeSubWithPostProcessor",
  "postProcessor": [Function],
  "type": "onCreate",
}
`;

exports[`Subscription postProcessor should pass through postProcessor if it is declared: withPostProcessorEnd 1`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`Subscription postProcessor should pass through postProcessor if it is declared: withPostProcessorResult 1`] = `
Object {
  "done": false,
  "value": Object {
    "data": Object {
      "someSubWithPostProcessorSomeTestsEntityName": Object {
        "someTestsEntityName": Object {
          "anotherAttribute": "earth",
        },
      },
    },
  },
}
`;

exports[`Subscription postProcessor should pass through postProcessor if it is declared: withoutPostProcessorEnd 1`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`Subscription postProcessor should pass through postProcessor if it is declared: withoutPostProcessorResult 1`] = `
Object {
  "done": false,
  "value": Object {
    "data": Object {
      "someSubWithPostProcessorSomeTestsEntityName": Object {
        "someTestsEntityName": Object {
          "anotherAttribute": "world",
        },
      },
    },
  },
}
`;

exports[`Subscription preProcessor should pass through preProcessor if it is declared 1`] = `
Subscription {
  "attributes": Array [
    "someAttribute",
  ],
  "delimiter": "/",
  "description": "build item",
  "ignoreRequired": true,
  "isTypeCreate": true,
  "name": "SomeSubWithPreProcessor",
  "preProcessor": [Function],
  "type": "onCreate",
}
`;

exports[`Subscription preProcessor should pass through preProcessor if it is declared: withPreProcessorEnd 1`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`Subscription preProcessor should pass through preProcessor if it is declared: withPreProcessorResult 1`] = `
Object {
  "done": false,
  "value": Object {
    "data": Object {
      "someSubWithPreProcessorSomeTestsEntityName": Object {
        "someTestsEntityName": Object {
          "anotherAttribute": "world",
        },
      },
    },
  },
}
`;

exports[`Subscription preProcessor should pass through preProcessor if it is declared: withoutPreProcessorEnd 1`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`Subscription preProcessor should pass through preProcessor if it is declared: withoutPreProcessorResult 1`] = `
Object {
  "done": false,
  "value": Object {
    "data": Object {
      "someSubWithPreProcessorSomeTestsEntityName": Object {
        "someTestsEntityName": Object {
          "anotherAttribute": "world",
        },
      },
    },
  },
}
`;

exports[`Subscription processEntitySubscriptions should throw if provided with an invalid list of subscriptions 1`] = `"Entity 'SomeEntityName' subscriptions definition needs to be an array of subscriptions"`;

exports[`Subscription processEntitySubscriptions should throw if provided with an invalid subscription 1`] = `"Invalid subscription definition for entity 'SomeEntityName' at position '0'"`;

exports[`Subscription processEntitySubscriptions should throw if required attribute is missing in CREATE type subscriptions 1`] = `"Missing required attributes in subscription 'SomeEntityName.onBuild' need to have a defaultValue() function: [ neededAttribute ]"`;

exports[`Subscription processEntitySubscriptions should throw if unknown attributes are used 1`] = `"Cannot use attribute 'SomeEntityName.doesNotExist' in subscription 'SomeEntityName.onBuild' as it does not exist"`;

exports[`Subscription processEntitySubscriptions should throw on duplicate subscription names 1`] = `"Duplicate subscription name 'onBuild' found in 'SomeEntityName'"`;

exports[`Subscription should have a description 1`] = `"Missing description for subscription 'example'"`;

exports[`Subscription should have a list of unique attribute names 1`] = `"Subscription 'SomeEntityName.example' needs to have a list of unique attribute names"`;

exports[`Subscription should have a list of valid attribute names 1`] = `"Subscription 'SomeEntityName.example' needs to have a list of attribute names"`;

exports[`Subscription should have a name 1`] = `"Missing subscription name"`;

exports[`Subscription should have a type 1`] = `"Missing type for subscription 'example'"`;

exports[`Subscription should have a valid postProcessor function 1`] = `"postProcessor of subscription 'example' needs to be a valid function"`;

exports[`Subscription should have a valid preProcessor function 1`] = `"preProcessor of subscription 'example' needs to be a valid function"`;

exports[`Subscription should have a valid type 1`] = `"Unknown subscription type '12346' used, try one of these: 'onCreate, onUpdate, onDelete'"`;
